File: pymlac/PtrPtp.py
#!/usr/bin/python

"""
Emulate the imlac Paper Tape Reader/Punch (PTR/PTP).

We take some pains to closely emulate the *real* device, even
including misuse, such as no tape mounted.  This means we must tell
the PTR/PTP object how many CPU cycles have passed (tick()).
"""

import struct

from Globals import *


class PtrPtp(object):

    # number of chars per second we want
    PtrCharsPerSecond = 300
    PtpCharsPerSecond = 30

    # duty cycle for PTR is 30% ready and 70% not ready
    PtrReadyCycles = int((CYCLES_PER_SECOND / PtrCharsPerSecond) / 0.7) / 25
    PtrNotReadyCycles = int((CYCLES_PER_SECOND / PtrCharsPerSecond) / 0.3) / 25
    PtpNotReadyCycles = int(CYCLES_PER_SECOND / PtpCharsPerSecond)

    # no tape in reader, return 0377 (all holes see light)
    PtrEOF = 0377

    # module-level state variables
    device_use = None
    device_motor_on = False
    device_open_file = None
    device_filename = None
    device_cycle_count = 0
    device_ready = False

    ptr_at_eof = True
    ptr_value = PtrEOF

    # status values
    InUsePTR = 'PTR'
    InUsePTP = 'PTP'


    def __init__(self):
        """Initialize the papertape device."""

        self.reset()

    def reset(self):
        """Reset device to a known state."""

        # general device
        self.device_use = None
        self.device_motor_on = False
        self.device_open_file = None
        self.device_filename = None
        self.device_open_file = None
        self.device_filename = None
        self.device_ready = False

        # reader-specific
        self.ptr_at_eof = True
        self.ptr_value = self.PtrEOF

    def ready(self):
        """Test if device is ready."""

        return self.device_ready

    ###############
    # Interface routines for reader
    ###############

    def ptr_mount(self, fname):
        """Mount papertape file on the reader device."""

        if self.device_use == self.InUsePTP:
            raise RuntimeError("ptr_mount: Can't mount PTR file, being used as PTP")

        self.device_use = self.InUsePTR
        self.device_motor_on = False
        self.device_filename = fname
        self.device_open_file = open(self.device_filename, 'rb')
        self.device_ready = False
        self.device_cycle_count = self.PtrNotReadyCycles
        self.ptr_at_eof = False
        self.ptr_value = None

    def ptr_dismount(self):
        """Dismount a papertape file."""

        if self.device_use == self.InUsePTP:
            raise RuntimeError("ptr_dismount: Can't dismount PTR file, being used as PTP")

        self.device_use = None
        self.device_motor_on = False
        self.device_ready = False
        if self.device_filename:
            self.device_open_file.close()
            self.device_open_file = None
        self.device_filename = None
        self.ptr_at_eof = True
        self.ptr_value = self.PtrEOF

    def start(self):
        """Turn papertape reader motor on."""

        if self.device_use == self.InUsePTP:
            raise RuntimeError("start: Can't start PTR motor, being used as PTP")

        self.device_use = self.InUsePTR
        self.device_motor_on = True
        self.device_ready = False
        self.device_cycle_count = self.PtrReadyCycles

    def stop(self):
        """Stop reader motor."""

        if self.device_use == self.InUsePTP:
            raise RuntimeError("stop: Can't stop PTR motor, being used as PTP")

        self.device_motor_on = False
        self.device_ready = False
        self.device_cycle_count = self.PtrReadyCycles

    def read(self):
        """Read papertape value."""

        if self.device_use == self.InUsePTP:
            raise RuntimeError("ptr_read: Can't read PTR, being used as PTP")

        return self.ptr_value

    def ptr_eof(self):
        """Return reader EOF status."""

        if self.device_use == self.InUsePTP:
            raise RuntimeError("ptr_eof: Can't read PTR status, being used as PTP")

        return self.ptr_at_eof

    def ptr_tick(self, cycles):
        """Called to push PTR state along.

        cycles  number of cycles passed since last tick
        """

        if self.device_use != self.InUsePTR:
            return

        # if end of tape or motor off, do nothing, state remains unchanged
        if self.ptr_at_eof or not self.device_motor_on:
            return

        self.device_cycle_count -= cycles

        if self.device_cycle_count <= 0:
            if self.device_ready:
                self.device_ready = False
                self.device_cycle_count += self.PtrNotReadyCycles
            else:
                self.device_ready = True
                self.device_cycle_count += self.PtrReadyCycles
                self.ptr_value = self.device_open_file.read(1)
                if len(self.ptr_value) < 1:
                    # EOF on input file, pretend end of tape
                    self.ptr_at_eof = True
                    self.ptr_value = self.PtrEOF
                else:
                    self.ptr_value = ord(self.ptr_value)

    ###############
    # Interface routines for punch
    ###############

    def ptp_mount(self, fname):
        """Mount papertape file on the punch device."""

        # check usage of the device
        if self.device_use == self.InUsePTR:
            raise RuntimeError("ptp_mount: Can't mount PTP file, being used as PTR")

        self.device_use = self.InUsePTP
        self.device_motor_on = False
        self.device_ready = False
        self.device_cycle_count = self.PtrNotReadyCycles
        self.ptp_filename = fname
        self.device_open_file = open(self.ptp_filename, 'wb')
        self.ptp_at_eof = False
        self.ptp_value = None

    def ptp_dismount(self):
        """Dismount a papertape punch file."""

        if self.device_use == self.InUsePTR:
            raise RuntimeError("ptp_dismount: Can't dismount PTP file, being used as PTR")

        self.device_motor_on = False
        self.device_ready = False
        if self.ptp_filename:
            self.device_open_file.close()
            self.device_open_file = None
        self.ptp_filename = None
        self.ptp_at_eof = True
        self.ptp_value = self.PtrEOF

    def punch(self, value):
        """Write byte to papertape file.

        value  8 bit value to punch
        """

        if self.device_use == self.InUsePTR:
            raise RuntimeError("punch: Can't punch PTP file, being used as PTR")

        self.device_ready = False
        self.device_cycle_count = self.PtpNotReadyCycles
        self.device_open_file.write(struct.pack('1B', value))

    def ptp_tick(self, cycles):
        """Called to push PTP state along.

        cycles  number of cycles passed since last tick
        """

        if self.device_use != self.InUsePTP:
            return

        self.device_cycle_count -= cycles
        if self.device_cycle_count <= 0:
            self.device_ready = True
            self.device_cycle_count += self.PtrNotReadyCycles

File: vimlac/ptrptp.c
/*
 * Implementation for the vimlac PTR/PTP (papertape reader/punch).
 */

#include "vimlac.h"
#include "ptrptp.h"
#include "log.h"


/*****
 * constants for the PTR device
 *
 * The device reads at 300 chars/second, so we work out how many
 * machine cycles data is ready/notready at a 30%/70% ready cycle.
 ******/

#define PTR_CHARS_PER_SECOND        300
#define PTR_CYCLES_PER_CHAR         (CPU_HERZ / PTR_CHARS_PER_SECOND)
#define PTR_READY_CYCLES            (int)((3 * PTR_CYCLES_PER_CHAR) / 10)
#define PTR_NOT_PTR_READY_CYCLES    (int)((7 * PTR_CYCLES_PER_CHAR) / 10)

#define PTR_EOF                     0377

/*****
 * constants for the PTP device
 *
 * The device punches at 30 chars/second, so we work out how many
 * machine cycles device is not ready after punch starts.
 ******/

#define PTP_CHARS_PER_SECOND    30
#define PTP_NOT_READY_CYCLES    (int) (CPU_HERZ / PTP_CHARS_PER_SECOND)

// define some use values
#define InUsePTR "PTR"
#define InUsePTP "PTP"


/*****
 * State variables for the general device
 ******/

static bool device_ready = false;
static char *device_use = NULL;         // NULL, InUsePTR or InUsePTP
static bool device_motor_on = false;
static FILE *device_open_file = NULL;
static char *device_filename = NULL;
static long device_cycle_count = 0;

/*****
 * Specific state variables for the PTR device
 ******/

static bool ptr_at_eof = true;
static BYTE ptr_value = PTR_EOF;


void ptrptp_reset(void)
{
    device_ready = false;
    device_use = NULL;
    device_motor_on = false;
    device_open_file = NULL;
    device_filename = NULL;
    device_cycle_count = 0;
}


int
ptr_mount(char *fname)
{
    if (STREQ(device_use, InUsePTP))
       error("ptr_mount: Can't mount PTR file, being used as PTP");

    device_filename = fname;
    device_open_file = fopen(fname, "rb");
    if (device_open_file == NULL)
    {
        ptr_dismount();
        error("ptr_mount: can't mount file %s", fname);
    }
    device_motor_on = false;
    device_ready = false;
    ptr_at_eof = false;
    ptr_value = PTR_EOF;
    device_cycle_count = PTR_NOT_PTR_READY_CYCLES;
    device_use = InUsePTR;

    return 0;
}


void
ptr_dismount(void)
{
    if (STREQ(device_use, InUsePTP))
       error("ptr_mount: Can't dismount PTR file, being used as PTP");

    if (device_open_file)
        fclose(device_open_file);
    device_filename = NULL;
    device_open_file = NULL;
    device_motor_on = false;
    device_ready = false;
    ptr_at_eof = true;
    ptr_value = PTR_EOF;
    device_use = NULL;
}


void
ptr_start(void)
{
    if (STREQ(device_use, InUsePTP))
       error("ptrptp_start: Can't start PTR motor, being used as PTP, device_use=%s", device_use);

    device_motor_on = true;
    device_ready = false;
    device_cycle_count = PTR_NOT_PTR_READY_CYCLES;
}


void
ptr_stop(void)
{
    if (STREQ(device_use, InUsePTP))
       error("ptr_stop: Can't stop PTR motor, being used as PTP");

    device_motor_on = false;
    device_cycle_count = PTR_NOT_PTR_READY_CYCLES;
    device_ready = false;
}


int
ptr_read(void)
{
    if (STREQ(device_use, InUsePTP))
        error("ptr_read: Can't read PTR, device being used as PTP");

    return ptr_value;
}


bool
ptr_ready(void)
{
    return device_ready;
}


void
ptr_tick(long cycles)
{
    // if not being used as PTR, do nothing
    if (STREQ(device_use, InUsePTP))
        return;

    /* if no state change */
    if (!device_motor_on || ptr_at_eof || device_open_file == NULL)
        return;

    /* tape in, motor on */
    device_cycle_count -= cycles;
    if (device_cycle_count <= 0L)
    {
        if (device_ready == true)
        {
            device_ready = false;
            device_cycle_count += PTR_NOT_PTR_READY_CYCLES;
        }
        else
        {
            device_ready = true;
            device_cycle_count += PTR_READY_CYCLES;
            if (fread(&ptr_value, sizeof(BYTE), 1, device_open_file) != 1)
            {   /* EOF on file */
                ptr_at_eof = true;
                ptr_value = PTR_EOF;
            }
        }
    }
}

// MID

int
ptp_mount(char *fname)
{
    if (STREQ(device_use, InUsePTR))
        error("ptp_mount: Can't mount PTP, device being used as '%s'", device_use);

    device_use = InUsePTP;
    device_filename = fname;
    device_open_file = fopen(fname, "wb");
    if (device_open_file == NULL)
    {
        ptp_dismount();
        error("ptp_mount: Can't mount file %s", fname);
    }
    device_motor_on = false;
    device_ready = false;
    device_cycle_count = PTP_NOT_READY_CYCLES;

    return 0;
}


void
ptp_dismount(void)
{
    if (STREQ(device_use, InUsePTR))
       error("ptp_dismount: Can't dismount PTP, device being used as PTR");

    if (device_open_file)
        if (fclose(device_open_file) != 0)
    device_filename = NULL;
    device_open_file = NULL;
    device_motor_on = false;
    device_ready = true;
}


void
ptp_start(void)
{
    if (STREQ(device_use, InUsePTR))
       error("ptp_dismount: Can't start PTP, device being used as PTR");

    device_motor_on = true;
    device_ready = false;
    device_cycle_count = PTP_NOT_READY_CYCLES;
}


void
ptp_stop(void)
{
    if (STREQ(device_use, InUsePTR))
       error("ptp_stop: Can't stop PTP, device being used as PTR");

    device_motor_on = false;
    device_cycle_count = PTP_NOT_READY_CYCLES;
}


void
ptp_punch(BYTE value)
{
    if (STREQ(device_use, InUsePTR))
       error("ptp_punch: Can't punch PTP, device being used as PTR");

    if (device_motor_on && device_open_file != NULL)
    {
        putc(value, device_open_file);
        device_cycle_count = PTP_NOT_READY_CYCLES;
    }
}


bool
ptp_ready(void)
{
    return device_ready;
}


void
ptp_tick(long cycles)
{
    if (STREQ(device_use, InUsePTR))
       return;

    /* if no mounted file */
    if (device_open_file == NULL)
        return;

    /* tape in */
    device_cycle_count -= cycles;
    if (device_cycle_count <= 0L)
    {
        if (!device_ready)
            device_ready = true;
        device_cycle_count = 0;
    }
}
